cpp/pipelines/add_authority_wikidata_ids.cc

```diff
@@ -0,0 +1,195 @@
+        while (std::getline(file, line)) {
+            if (StringUtil::StartsWith(line, "Name:") and StringUtil::Contains(line, "GND:") and StringUtil::Contains(line, "Wikidata:")) {
+                act_gnd = line.substr(line.find("GND:") + 4);
+                act_gnd = act_gnd.substr(0, act_gnd.find("Wikidata:"));
+                act_wiki = line.substr(line.find("Wikidata:") + 9);
+                gnd_to_wikidataid->emplace(trim(act_gnd), trim(act_wiki));
+            }
+        }
+        file.close();
+    }
+    else 
+        LOG_ERROR("input or output files could not be opened");
+}
+
+
+int Main(int argc, char * argv[]) {
+
+     if (argc != 4)
+        Usage();
+
+    const std::string marc_input_filename_or_create_flag(argv[1]);
+    const std::string marc_output_filename_or_dnb_input(argv[2]);
+    const std::string mapping_txt_filename(argv[3]);
+
+    if (marc_input_filename_or_create_flag == "--create_mapping_file") {
+        //e.g. "/..../authorities-person_lds_20210613.jsonld" and /usr/local/ub_tools/cpp/data/gnd_to_wiki.txt
+        ParseDataDnbFile(marc_output_filename_or_dnb_input, mapping_txt_filename); 
+        return EXIT_SUCCESS;
+    }
+
+    std::unordered_map<std::string, std::string> gnd_to_wikidataid;
+    ParseGndWikidataMappingFile(mapping_txt_filename, &gnd_to_wikidataid);
+    
+    std::unique_ptr<MARC::Reader> marc_reader(MARC::Reader::Factory(marc_input_filename_or_create_flag, MARC::FileType::BINARY));
+    std::unique_ptr<MARC::Writer> marc_writer(MARC::Writer::Factory(marc_output_filename_or_dnb_input, MARC::FileType::BINARY));
+
+    if (unlikely(marc_input_filename_or_create_flag == marc_output_filename_or_dnb_input))
+        LOG_ERROR("Norm data input file name equals output file name!");
+
+    while (MARC::Record record = marc_reader.get()->read()) {
+        // 035|a (DE-588)118562215
+        std::string record_gnd;
+        std::string wikidata_id;
+
+        auto field_035(record.getFirstField("035"));
+        while (field_035 != record.end() and field_035->getTag() == "035") {
+            if (not field_035->getFirstSubfieldWithCode('a').empty() and
+                field_035->getFirstSubfieldWithCode('a').find("(DE-588)") != std::string::npos)
+            {
+                record_gnd = field_035->getFirstSubfieldWithCode('a');
+                record_gnd = record_gnd.substr(record_gnd.find("(DE-588)") + 8);
+                record_gnd = trim(record_gnd);
+                break;
+            }
+            ++field_035;
+        }
+        //record lookup
+        if (not record_gnd.empty()) {
+            auto gnd_to_wikidataid_iter = gnd_to_wikidataid.find(record_gnd);
+            if (gnd_to_wikidataid_iter != gnd_to_wikidataid.end()) {
+                wikidata_id = gnd_to_wikidataid_iter->second;
+                //std::cout << "Match: " << record_gnd << " --- " << wikidata_id << "\n";
+            }
+        }
+        //record write wikidata id (if not already there)
```

Comment:
It might be useful to move this whole block into MARC.h as well, similar to existing functions like Record::getDOIs()

Reply:
Also, performance might be better if you check wikidata_existing and skipping the record before executing the lookup against the gnd_to_wikidataid mapping file.